File: packages/asdl-objectives/src/asdl_objectives/exec/list_candidates.py
"""``objective exec list-candidates`` fast autocomplete candidate inventory."""

from __future__ import annotations

import click

from asdl_core.clinkr.exit import ClinkrExit
from asdl_core.clinkr.models import ClinkrModel
from asdl_core.clinkr.operation import clinkr_operation
from asdl_objectives.context import ObjectiveCliUnavailable, load_objective_context
from asdl_objectives.list_status import matches_status_filter
from asdl_objectives.objective_storage import FilesystemObjectiveStorage, ObjectiveRecordStatus


class ObjectiveCandidateRecord(ClinkrModel):
    slug: str
    status: ObjectiveRecordStatus


class ObjectiveListCandidatesRequest(ClinkrModel):
    pass


class ObjectiveListCandidatesResult(ClinkrModel):
    records: tuple[ObjectiveCandidateRecord, ...]


def render_list_candidates(result: ObjectiveListCandidatesResult) -> None:
    for record in result.records:
        click.echo(f"{record.slug}\t{record.status}")


@clinkr_operation(
    name="list-candidates",
    help="List active Objective slug candidates for shell and agent autocomplete.",
    human_renderer=render_list_candidates,
)
def run_list_candidates(
    ctx: click.Context,
    request: ObjectiveListCandidatesRequest,
) -> ClinkrExit[ObjectiveListCandidatesResult]:
    del request
    objective_ctx = load_objective_context(ctx)
    if isinstance(objective_ctx, ObjectiveCliUnavailable):
        return ClinkrExit.failure(error_type="not_in_repo", message=objective_ctx.message)

    inventory = FilesystemObjectiveStorage(objective_ctx.repo_root).checkout_inventory()
    records = tuple(
        ObjectiveCandidateRecord(slug=record.slug, status=record.status)
        for record in inventory.records
        if matches_status_filter(record.status, "active")
    )
    return ClinkrExit.ok(ObjectiveListCandidatesResult(records=records))

File: ts/packages/objective/src/operations/list-candidates.ts
import { failure, ok, type ClinkrExit, type LegacyMachineOutput } from "@asdl/clinkr";
import { z } from "zod";

import type { ObjectiveCliContext } from "../context.ts";
import { legacyMachine } from "./legacy-machine.ts";
import { matchesStatusFilter } from "./list-objectives.ts";

export const listCandidatesRequestSchema = z.object({});

export const objectiveCandidateRecordSchema = z.object({
	slug: z.string(),
	status: z.enum(["open", "closed"]),
});

export const listCandidatesResultSchema = z.object({
	records: z.array(objectiveCandidateRecordSchema),
});

export type ObjectiveCandidateRecord = z.infer<typeof objectiveCandidateRecordSchema>;
export type ListCandidatesResult = z.infer<typeof listCandidatesResultSchema>;
export type ListCandidatesRequest = z.infer<typeof listCandidatesRequestSchema>;

export async function runListCandidates(
	ctx: ObjectiveCliContext,
	request: ListCandidatesRequest,
): Promise<ClinkrExit<ListCandidatesResult>> {
	void request;
	const inventory = await ctx.storage.checkoutInventory();
	if (!inventory.ok) return failure(inventory.error.code, inventory.error.message);

	return ok({
		records: inventory.value.records
			.filter((record) => matchesStatusFilter(record.status, "active"))
			.map((record) => ({ slug: record.slug, status: record.status })),
	});
}

export function renderListCandidates(result: ListCandidatesResult): string {
	return result.records.map((record) => `${record.slug}\t${record.status}`).join("\n");
}

export function legacyListCandidatesMachine(exit: ClinkrExit<ListCandidatesResult>): LegacyMachineOutput {
	return legacyMachine(exit);
}
